File: hrm/custom_script/employee/employee.py
# -*- coding: utf-8 -*-
# Copyright (c) 2018, Digitalprizm and contributors
# For license information, please see license.txt

from __future__ import unicode_literals
import frappe
from hrm.hrm.doctype.vacation_rejoining.vacation_rejoining import vacation_rejoining_edit
from frappe.utils import getdate
from datetime import datetime, timedelta

@frappe.whitelist()
def validate(doc, method):
	if not doc.rejoining_date: doc.vacation_opening_balance = 0

	validate_rejoining_date(doc)
	validate_airline(doc)
	validate_onupdate(doc)

@frappe.whitelist()
def onload(doc, method):
	validate_future_vacation(doc)

def validate_onupdate(doc):
	if doc.is_new() == None and doc.get("__onload"):
		validate_future_vacation(doc)
		date_change = find_date_change(doc)
		if doc.get("__onload")["read_only"] == 1 and date_change == 1:
			frappe.throw("Joining Date OR Rejoining Date Cannot be Updated")

def validate_future_vacation(doc):
	if doc.is_new() == None:
		result_parm = vacation_rejoining_edit(doc.employee)
		modif_parm = validate_rule_modification(doc)

		doc.get("__onload")["read_only"] = modif_parm if result_parm == 1 else 1

def validate_rule_modification(doc):
	modif_select = """SELECT *
	FROM `tabVacation Rule Modification`
	WHERE docstatus < 2
	AND employee = '{0}'
	""".format(doc.employee)
	modif_result = frappe.db.sql(modif_select,as_list = True)

	return 1 if modif_result else 0

def find_date_change(doc):
	modif_select = """SELECT *
	FROM `tabEmployee`
	WHERE name = '{0}'
	AND( date_of_joining != '{1}'
	OR rejoining_date != '{2}'
	OR vacation_opening_balance != '{3}')
	""".format(doc.employee, doc.date_of_joining, doc.rejoining_date, doc.vacation_opening_balance)
	modif_select = frappe.db.sql(modif_select, as_list = True)

	return 1 if modif_select else 0

def validate_airline(doc):
	if doc.eligible_for_airline_ticket == 1:
		if (doc.number_of_trips or 0) <= 0:
			frappe.throw("Number Of Trips Should not be Zero")

		if (doc.year or 0) <= 0:
			frappe.throw("Per Year Should not be Zero")

		if (doc.eligible_head_count_including_self or 0) <= 0:
			frappe.throw("Eligible head count including self Should not be Zero")

		if doc.mode_of_reimbursement == "Cash" and (doc.eligible_cash or 0) <= 0:
			frappe.throw("Eligible cash Should not be Zero")

def validate_rejoining_date(doc):
	if doc.rejoining_date and getdate(str(doc.rejoining_date)) < getdate(str(doc.date_of_joining)):
		frappe.throw("Vacation Rejoining Date Cannot be Less then Date of Joining")


@frappe.whitelist()
def fetch_child(templete, date_of_joining):
	list_data = []
	date_jo = datetime.strptime(date_of_joining, '%Y-%m-%d')
	sql = """select days, document, responsible_persons
		from `tabDocument List`
		where parent = '{}'""".format(templete)
	data = frappe.db.sql(sql, as_dict=True)

	for i in data:
		days = i['days']
		future = date_jo + timedelta(days=int(days))
		dat = datetime.strftime(future, '%Y-%m-%d')
		document = i['document']
		list_data.append({"document": document,
			"date": dat})
	
	return list_data

File: hrm/custom_script/employee/employee.js
frappe.ui.form.on("Employee", {
	setup: function(frm) {
		frm.trigger('filter_vacation_rule');

		frm.set_query('ot_rule', function(doc) {
			return {
				filters: {
					docstatus: 1
				}
			}
		});
	},

	refresh: function(frm) {
		frm.trigger('set_read_only');
	},

	eligible_for_airline_ticket: function(frm) {
		var eligible = frm.doc.eligible_for_airline_ticket;

		frm.set_df_property('class', 'reqd', eligible == 1 ? 1 : 0);
		frm.set_df_property('mode_of_reimbursement', 'reqd', eligible == 1 ? 1 : 0);
		frm.set_df_property('origin_airport', 'reqd',  eligible == 1 ? 1 : 0);
		frm.set_df_property('destination_airport', 'reqd', eligible == 1 ? 1 : 0);
		frm.set_df_property('number_of_trips', 'reqd', eligible == 1 ? 1 : 0);
		frm.set_df_property('year', 'reqd', eligible == 1 ? 1 : 0);
		frm.set_df_property('eligible_head_count_including_self', 'reqd', eligible == 1 ? 1 : 0);
		frm.set_value("number_of_trips", eligible == 1 ? 1 : 0);
		frm.set_value("year", eligible == 1 ? 1 : 0);

		if(frm.doc.eligible_for_airline_ticket == 0) {
			frm.set_value("class", undefined);
			frm.set_value("mode_of_reimbursement", undefined);
			frm.set_value("origin_airport", undefined);
			frm.set_value("destination_airport", undefined);
		}
	},

	mode_of_reimbursement: function(frm) {
		frm.set_df_property('eligible_cash', 'reqd', frm.doc.mode_of_reimbursement == "Cash" ? 1 : 0);
		if(!frm.doc.mode_of_reimbursement == "Cash") {
			frm.set_value("eligible_cash", 0);
		}
	},

	filter_vacation_rule: function(frm) {
		frm.fields_dict['vacation_rule'].get_query = function(doc) {
			return {
				filters: {
					docstatus: 1
				}
			}
		}
	},

	rejoining_date: function(frm) {
		if(frm.doc.rejoining_date && frm.doc.rejoining_date < frm.doc.date_of_joining) {
			frappe.throw("Vacation Rejoining Date Cannot be Less then Date of Joining");
		}
	},

	set_read_only: function(frm) {
		// set rejoining and vacation rule readonly if there is any vaction for future
		var result = frm.doc.__onload ? frm.doc.__onload.read_only : 0;
		frm.set_df_property('rejoining_date', 'read_only', result);
		frm.set_df_property('vacation_rule', 'read_only', result);
		frm.set_df_property('date_of_joining', 'read_only', result);
		frm.set_df_property('vacation_opening_balance', 'read_only', result);
	},

	first_name: function(frm) {
		if(frm.doc.firstname == undefined) {
			frm.set_value('employee_name', undefined);
		}
	},

	template: function(frm) {
		if (!frm.doc.template || !frm.doc.date_of_joining) {
			frm.set_value('documents', []);
			frm.refresh_field('documents');
			return
		}
		
		frappe.call({
			method: "hrm.doctype_triggers.hr.employee.employee.fetch_child",
			args:
			{
				"templete": frm.doc.template,
				"date_of_joining": frm.doc.date_of_joining,
			},
			callback: function (r) {
				frm.set_value('documents', r.message);
				frm.refresh_field('documents');
			}
		});
	}
});